File: frontend/src/types/gift.ts
/**
 * Gift domain types — aligned with BE /api/gifts contract.
 */

export type GiftStatus = "CREATED" | "OPENED" | "SAVED";

/** 자산 포맷 — 듀얼 포맷 분기. */
export type AssetKind = "TRACKLIST" | "MOOD_VIDEO";

/** 선물 계기(5값). */
export type Occasion =
  | "BIRTHDAY"
  | "ANNIVERSARY"
  | "COMFORT"
  | "CELEBRATION"
  | "JUST_BECAUSE";

/** 계기 → 한국어 라벨/이모지 (UI). */
export const OCCASION_LABELS: Record<Occasion, string> = {
  BIRTHDAY: "🎂 생일",
  ANNIVERSARY: "💞 기념일",
  COMFORT: "🌙 위로",
  CELEBRATION: "🎉 축하",
  JUST_BECAUSE: "💌 그냥",
};

export interface GiftSender {
  handle: string;
  displayName: string;
  avatarUrl?: string;
}

export interface GiftAsset {
  id: string;
  title: string;
  coverUrl?: string;
  tracks: GiftTrack[];
  /** 듀얼 포맷 — 뷰가 재생 컴포넌트를 이걸로 분기. */
  assetKind: AssetKind;
  moodVideoId?: string | null;
  moodChannelName?: string | null;
  moodTrackListText?: string | null;
}

export interface GiftTrack {
  id: string;
  name: string;
  artist: string;
  durationMs?: number;
  isrc?: string;
  spotifyId?: string;
  appleMusicId?: string;
  youtubeVideoId?: string | null;
}

/** POST /api/gifts response */
export interface GiftCreated {
  token: string;
  url: string;
}

/** GET /api/gifts/{token} response */
export interface GiftView {
  token: string;
  status: GiftStatus;
  message: string;
  wrapSkin: string;
  /** 헌정 — "지영에게". message 와 분리돼 봉투/언박싱/OG 에 단독 노출. */
  dedicationTo?: string | null;
  occasion?: Occasion | null;
  sender: GiftSender;
  asset: GiftAsset;
}

/** GET /api/gifts/received|sent 목록 항목(요약) */
export interface GiftSummary {
  token: string;
  status: GiftStatus;
  message: string;
  wrapSkin: string;
  dedicationTo?: string | null;
  occasion?: Occasion | null;
  sender: GiftSender;
  assetTitle: string;
  assetCoverUrl?: string;
  trackCount: number;
  assetKind: AssetKind;
  createdAt: string;
}

/** Nocturne 포장 스킨 정의 */
export interface WrapSkin {
  key: string;
  label: string;
  /** Tailwind/CSS gradient class or inline style value */
  gradient: string;
  accentColor: string;
}

export const WRAP_SKINS: WrapSkin[] = [
  {
    key: "nocturne-violet",
    label: "미드나잇 바이올렛",
    gradient: "linear-gradient(135deg, #2d1b69 0%, #11074a 100%)",
    accentColor: "#a78bfa",
  },
  {
    key: "nocturne-rose",
    label: "딥 로즈",
    gradient: "linear-gradient(135deg, #4a1128 0%, #1a0a1a 100%)",
    accentColor: "#f472b6",
  },
  {
    key: "nocturne-teal",
    label: "오션 틸",
    gradient: "linear-gradient(135deg, #0d3d3d 0%, #071a1a 100%)",
    accentColor: "#2dd4bf",
  },
  {
    key: "nocturne-gold",
    label: "앰버 골드",
    gradient: "linear-gradient(135deg, #3d2a00 0%, #1a1000 100%)",
    accentColor: "#fbbf24",
  },
];

File: backend/src/main/kotlin/com/plshare/backend/domain/gift/model/Gift.kt
package com.plshare.backend.domain.gift.model

import jakarta.persistence.*
import java.time.LocalDateTime
import java.util.*

/**
 * 감성 선물 엔티티.
 *
 * 발신자(senderId)가 자신의 자산(assetId)에 메시지와 포장 스킨을 입혀
 * 고유 토큰으로 수신자에게 링크를 전달하는 도메인 객체.
 *
 * 수명 주기:
 * - CREATED  : 선물 생성됨, 아직 열리지 않음.
 * - OPENED   : 수신자가 언박싱 진입 — openedAt 기록(멱등).
 * - SAVED    : 수신자가 라이브러리에 저장 — savedByUserId 기록.
 *
 * v1 제약: 저장(SAVED) 시 자산 복제 없이 수령 기록만 남긴다.
 * 향후 자산 소유권 모델(ownerId) 도입 시 savedByUserId를 기준으로
 * 수령자 라이브러리로 Asset을 복제하는 흐름 추가 필요.
 */
@Entity
@Table(
    name = "gifts",
    indexes = [
        Index(name = "idx_gifts_token", columnList = "token", unique = true),
        Index(name = "idx_gifts_sender_id", columnList = "sender_id"),
    ]
)
class Gift(
    @Id
    val id: UUID = UUID.randomUUID(),

    /** 선물 발신자의 User UUID. */
    @Column(name = "sender_id", nullable = false)
    val senderId: UUID,

    /** 포장 대상 자산 UUID. */
    @Column(name = "asset_id", nullable = false)
    val assetId: UUID,

    /** 발신자가 작성한 감성 메시지 (최대 3000자, 편지 길이). */
    @Column(name = "message", nullable = false, length = 3000)
    val message: String,

    /**
     * 포장 스킨 키.
     * 예: "nocturne-violet", "nocturne-rose", "nocturne-teal", "nocturne-gold"
     */
    @Column(name = "wrap_skin", nullable = false, length = 64)
    val wrapSkin: String,

    /**
     * 공개 링크용 고유 토큰 (16자 hex).
     * share token 패턴 차용: UUID.randomUUID().toString().replace("-","").take(16).
     * DB UNIQUE 인덱스로 충돌 방지.
     */
    @Column(name = "token", nullable = false, unique = true, length = 32)
    val token: String,

    /**
     * 헌정 — 받는 사람 이름/애칭(최대 40자). message 본문과 분리해 봉투 겉면·언박싱 첫 화면·OG 카드에
     * 단독 렌더한다. 익명 토큰 수신자를 '이 선물의 주인공'으로 못박는 키프세이크의 출발점.
     */
    @Column(name = "dedication_to", length = 40)
    var dedicationTo: String? = null,

    /** 계기 — '왜 주는가'. 포장 추천·OG 카피·회상 정렬에 재사용. 미선택(null) 허용. */
    @Enumerated(EnumType.STRING)
    @Column(name = "occasion", length = 20)
    var occasion: Occasion? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 16)
    var status: GiftStatus = GiftStatus.CREATED,

    @Column(name = "created_at", nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    /** 최초 언박싱 시각. open 호출 시 멱등하게 기록된다. */
    @Column(name = "opened_at")
    var openedAt: LocalDateTime? = null,

    /**
     * 저장한 수신자의 User UUID.
     * v1: 자산 복제 없이 수령 기록 전용.
     */
    @Column(name = "saved_by_user_id")
    var savedByUserId: UUID? = null,
)

enum class GiftStatus {
    CREATED,
    OPENED,
    SAVED,
}

/** 선물 계기 — 5값으로 압축(포장 추천·카피 분기 비용을 줄임). */
enum class Occasion {
    BIRTHDAY,
    ANNIVERSARY,
    COMFORT,
    CELEBRATION,
    JUST_BECAUSE,
}
